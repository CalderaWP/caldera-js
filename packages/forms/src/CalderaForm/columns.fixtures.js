import {
	checkboxFieldset,
	selectField,
	checkboxField,
	numberField,
	textField,
	emailField,
	radioField
} from './fields.fixtures';
import React from 'react';

export const formRowOne = {
	rowId: 'r1',
	columns: [
		{
			fields: [emailField],
			width: '1/2',
			columnId: '1a'
		},
		{
			fields: [textField],
			width: '1/4',
			columnId: '1b'
		}
	]
};

export const formRowTwo = {
	rowId: 'r2',
	columns: [
		{
			fields: [checkboxField, selectField],
			width: '1/2',
			columnId: 1
		},
		{
			fields: [radioField],
			width: '1/4',
			columnId: 2
		}
	]
};

export const formRows = [formRowOne, formRowTwo];

export const notFormRow = {
	rowId: 'notForm',
	columns: [
		{
			children: <div>Half Column</div>,
			width: '1/2',
			columnId: 1
		},
		{
			children: <div>Quarter Column</div>,
			width: '1/4',
			columnId: 2
		},
		{
			children: <div>Quarter Column 2</div>,
			width: '1/4',
			columnId: 3
		}
	]
};

export const notFormRows = [notFormRow];
