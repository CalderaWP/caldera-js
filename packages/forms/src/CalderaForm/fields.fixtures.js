const checkboxFieldset = {
	fieldType: 'checkboxes',
	label: 'Checkbox Field Set Label',
	fieldId: 'checkboxFieldSetLabel',
	description: 'Checkbox field set description',
	required: true,
	options: [
		{
			value: 1,
			label: 'One',
			id: 'opt-1'
		},
		{
			value: 2,
			label: 'Two',
			id: 'opt-2',
			description: 'The Second Option',
			attributes: {
				checked: true
			}
		},
		{
			value: 3,
			label: 'Three',
			id: 'opt-3'
		}
	]
};

const selectField = {
	fieldType: 'select',
	value: 1,
	label: 'Select Field Label',
	fieldId: 'selectFieldId',
	description: 'Select field description',
	required: true,
	attributes: {
		multiple: false
	},
	options: [
		{ value: 0, label: 'Zero' },
		{ value: 1, label: 'One' },
		{ value: 2, label: 'Two' }
	]
};

const radioField = {
	fieldType: 'radio',
	value: 1,
	label: 'Radio Field Label',
	fieldId: 'radioFieldLabel',
	description: 'Radio field description',
	required: true,
	attributes: {
		multiple: false
	},
	options: [
		{ value: 0, label: 'Zero' },
		{ value: 1, label: 'One' },
		{ value: 2, label: 'Two' }
	]
};

const checkboxField = {
	fieldType: 'checkbox',
	label: 'Checkbox Labe',
	fieldId: 'checkboxFieldLabel',
	description: 'Checkbox field description',
	required: true,
	attributes: {
		checked: true
	}
};

const numberField = {
	fieldType: 'number',
	value: 3,
	label: 'Number Field Label',
	fieldId: 'numberFieldId',
	description: 'Number field description',
	required: true,
	attributes: {
		min: 1,
		max: 10
	}
};

const textField = {
	fieldType: 'text',
	value: 'roy',
	label: 'First Name',
	fieldId: 'firstName',
	description: 'your first name',
	required: true
};

const emailField = {
	fieldType: 'email',
	value: 'emai@email.com',
	label: 'Email Field Label',
	fieldId: 'emailFieldId',
	description: 'Email field description',
	required: true,
	attributes: {
		multiple: true
	}
};

export {
	checkboxFieldset,
	selectField,
	checkboxField,
	numberField,
	textField,
	emailField,
	radioField
};
