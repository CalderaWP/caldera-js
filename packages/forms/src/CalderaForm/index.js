export { CalderaForm } from './CalderaForm';
export { CalderaGrid } from './CalderaGrid';
