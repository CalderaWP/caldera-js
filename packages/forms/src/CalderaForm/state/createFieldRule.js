/**
 * Create a conditional rule
 *
 * @param {string} test
 * @param {string} fieldId
 * @param {string|int|array} testValue
 * @return {function(*=): boolean}
 */
export const createFieldRule = (test,fieldId,testValue) => {
	const findFieldValue= (fieldId,fieldValues) => {
		return fieldValues.hasOwnProperty(fieldId)
			?fieldValues[fieldId]
			: null;
	};

	switch(test){
		case 'is':
			return (fieldValues) => {
				const value = findFieldValue(fieldId,fieldValues);
				return value == testValue;
			};
		case 'not':
			return (fieldValues) => {
				const value = findFieldValue(fieldId,fieldValues);
				return value !== testValue;
			}

	};
}
