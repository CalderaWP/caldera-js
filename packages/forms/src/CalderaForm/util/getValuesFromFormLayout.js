import {collectFieldValues} from '@calderawp/components';

/**
 * Given a Caldera Layout for a form, grab all field values
 *
 * @todo memoize
 *
 * @param {*} rows All rows of layout
 */
export const getValuesFromFormLayout = rows => {
	let values = {};
	rows.map(formRow => {
		const { columns } = formRow;
		columns.map(column => {
			const { fields } = column;
			values = {
				...values,
				...collectFieldValues(fields)
			};
		});
	});
	return values;
};
