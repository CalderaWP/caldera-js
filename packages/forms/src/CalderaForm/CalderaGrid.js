import PropTypes from 'prop-types';
import React, {createElement, Fragment} from 'react';
import classNames from 'classnames';
import {ConditionalState} from "./state/ConditionalState";
import {Column, Row, FieldArea} from '@calderawp/components';
/**
 *
 * @param rows
 * @param onAnyChange Function that runs when any supplied field value changes. Current field values are passed to callback.
 * @param onAnyBlur Function that runs when any field inside of this layout is blurred. Current field values are passed to callback.
 * @param {*} fieldValues Current field values used in layout. {fieldId: fieldValue}
 * @param {*} fieldErrors Current field errors. {fieldId: 'Invalid!'}
 * @param {*} fieldTouched
 * @param {function} setFieldValue
 * @return {*}
 * @constructor
 */
export const CalderaGrid = ({
								rows,
								onAnyChange,
								onAnyBlur,
								fieldValues,
								fieldErrors,
								fieldTouched,
								setFieldValue,
								conditionalState,
								applyConditionalRules
							}) => (
	<Fragment>
		{rows.map(row => {
			const {rowId, columns, render} = row;

			if (render) {
				return createElement(render, {
					...row,
					key: rowId
				});
			}

			return (
				<Row
					className={classNames('caldera-form-row')}
					key={rowId}
					id={rowId}
				>
					{columns.map(column => {
						const {
							padding,
							width,
							columnId,
							fields,
							render,
							key
						} = column;
						if (render) {
							return createElement(render, {
								...column,
								key: columnId
							});
						}
						return (
							<Column
								key={columnId}
								columnId={columnId}
								width={width}
								padding={padding}
							>
								{fields.map(field => {
									if (!field) {
										return;
									}
									const {fieldId, render, key} = field;
									field.value = fieldValues[fieldId];

									const _key = render ? key : fieldId;
									return (
										<FieldArea
											render={render}
											key={_key}
											field={field}
											onChange={newValue => {
												conditionalState.setValue(fieldId, newValue);
												applyConditionalRules();
												setFieldValue(
													fieldId,
													conditionalState.getValue(fieldId),
													true
												);
												onAnyChange(fieldValues);
											}}
											onBlur={onAnyBlur}
											fieldErrors={fieldErrors}
											fieldsTouch={fieldTouched}
										/>
									);
								})}
							</Column>
						);
					})}
				</Row>
			);
		})}
	</Fragment>
);

CalderaGrid.propTypes = {
	applyConditionalRules: PropTypes.func,
	conditionalState: PropTypes.instanceOf(ConditionalState),
	rows: PropTypes.array,
	onAnyChange: PropTypes.func,
	onAnyBlur: PropTypes.func,
	fieldValues: PropTypes.object,
	fieldErrors: PropTypes.object,
	fieldTouched: PropTypes.object,
	setFieldValue: PropTypes.func
};

const _noop = () => {
};
CalderaGrid.defaultProps = {
	onAnyChange: _noop,
	onAnyBlur: _noop,
	setFieldValue: _noop,
	fieldValues: {}
};
