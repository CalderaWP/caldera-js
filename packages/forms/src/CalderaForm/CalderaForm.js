import {ConditionalState} from './state/ConditionalState';
import PropTypes from 'prop-types';
import React, { isValidElement, createElement, Fragment, Component} from 'react';
import { Formik, Field, Form, ErrorMessage } from 'formik';
import {updateRows} from './util/updateRows';
import { collectFieldValues } from '@calderawp/components';
import {applyRuleToState} from './state/applyRule';

import { CalderaGrid } from './CalderaGrid';



export class CalderaForm extends Component {
	state = {
		formRows: [],
		initialValues: {},
		conditionalState: null
	};

	/**
	 * Run conditional logic and update state
	 */
	applyConditionalRules = () => {
		const {fields,rows,conditionals} = this.props.form;
		if( conditionals && conditionals.length ){
			const conditionalState = this.state.conditionalState ? this.state.conditionalState : new ConditionalState(collectFieldValues(fields ));
			conditionals.forEach( rule => {
				applyRuleToState(rule,conditionalState)
			});
			this.setState({
				formRows: updateRows(conditionalState,rows,fields),
				conditionalState
			});

		}

	};

	/**
	 * On mount, calculate initial state and rows
	 */
	componentDidMount =() =>{
		const {fields,rows} = this.props.form;
		const intialValues = collectFieldValues(fields );
		const conditionalState = this.state.conditionalState ? this.state.conditionalState : new ConditionalState(intialValues);
		const formRows = updateRows( conditionalState,rows,fields );
		this.setState({intialValues,formRows,conditionalState});
		this.applyConditionalRules();//initial hide/show logic
	};

	render(){
		const {onSubmit,onChange} = this.props;
		const {formRows,initialValues,conditionalState} = this.state;


		return (
			<div>
				<Formik
					initialValues={initialValues}
					onSubmit={onSubmit}
					render={({
								 errors,
								 status,
								 touched,
								 isSubmitting,
								 handleChange,
								 handleBlur,
								 setFieldValue,
								 values
							 }) => (
						<Form>
							<CalderaGrid
								applyConditionalRules={this.applyConditionalRules}
								conditionalState={conditionalState}
								rows={formRows}
								onAnyChange={onChange}
								onAnyBlur={handleBlur}
								fieldValues={values}
								setFieldValue={setFieldValue}
								fieldErrors={errors}
								fieldTouched={touched}
							/>
							<input
								type="submit"
								className={'caldera-forms-submit'}
								disabled={isSubmitting}
								value={'Click Button'}
							/>
						</Form>
					)}
				/>
			</div>
		);
	}

}

CalderaForm.propTypes = {
	form: PropTypes.object,
	onSubmit: PropTypes.func,
	onChange: PropTypes.func,
	onBlur: PropTypes.func
};

const noop = () => {};
CalderaForm.defaultProps = {
	onChange: noop,
	onBlur: noop
};
