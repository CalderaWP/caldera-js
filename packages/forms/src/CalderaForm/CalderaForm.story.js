import React from 'react';
import { storiesOf } from '@storybook/react';
import { CalderaForm } from './CalderaForm';
import { formRows } from './columns.fixtures';

import {collectFieldValues} from '@calderawp/components';
import {createFieldRule} from './state/createFieldRule';
import { getValuesFromFormLayout } from './util/getValuesFromFormLayout';
import {emailField, textField} from "./fields.fixtures";
import { FormTokenField } from '@wordpress/components';
import { withState } from '@wordpress/compose';


const form = {
	rows: [
		{
			rowId: 'r1',
			columns: [
				{
					fields: [emailField.fieldId],
					width: '1/2',
					columnId: '1aaaaa'
				},
				{
					fields: [textField.fieldId],
					width: '1/4',
					columnId: '1b'
				}
			]
		}
	],
	fields: [
		emailField,
		textField
	],
	conditionals :[
		{
			type: 'hide',
			rule: createFieldRule('is', emailField.fieldId, 'hide' ),
			fields: [
				textField.fieldId
			]
		},

	]
};

storiesOf('CalderaForm', module).add('Forms', () => (
	<CalderaForm
		form={form}
		onSubmit={(values, actions) => {
			setTimeout(() => {
				alert(JSON.stringify(values, null, 2));
				actions.setSubmitting(false);
			}, 1000);
		}}
	/>
));
