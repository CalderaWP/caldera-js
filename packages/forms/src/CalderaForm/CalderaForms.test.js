import {CalderaForm} from './CalderaForm';
import React, {Fragment} from 'react';
import renderer from 'react-test-renderer';
import {mount} from 'enzyme';
import {createFieldRule} from './state/createFieldRule';
import {formRows, formRowOne, checkboxField} from './columns.fixtures';
import {getValuesFromFormLayout} from './util/getValuesFromFormLayout';
import {emailField, textField} from './fields.fixtures';

import {Column,fieldAreaFactory,collectFieldValues} from '@calderawp/components';

describe( 'Factory import', () => {
	it( 'imports factory', () => {
		expect( typeof  fieldAreaFactory ).toBe( 'function' );
	});

	it( 'imports component', () => {
		expect( typeof Column ).toBe('function')
	});

	it( 'imports util function', () => {
		expect( typeof collectFieldValues ).toBe( 'function');
	})
});

describe('Caldera Forms', () => {
	let onChange = jest.fn();
	let onBlur = jest.fn();
	beforeEach(() => {
		onChange = jest.fn();
		onBlur = jest.fn();
	});
	const form = {
		rows: [
			{
				rowId: 'r1',
				columns: [
					{
						fields: [emailField.fieldId],
						width: '1/2',
						columnId: '1aaaaa'
					},
					{
						fields: [textField.fieldId],
						width: '1/4',
						columnId: '1b'
					}
				]
			}
		],
		fields: [
			emailField,
			textField
		]
	}

	it('Forms', () => {
		const component = renderer.create(
			<CalderaForm
				form={form}
				onBlur={onBlur}
				onChange={onChange}
			/>
		);
		expect(component.toJSON()).toMatchSnapshot();
	});

	it('Forms with rows assembled by id', () => {

		const component = mount(
			<CalderaForm
				form={form}
				onBlur={onBlur}
				onChange={onChange}
			/>
		);
		expect(component.find('#' + textField.fieldId).children().length).toBe(2);
		expect(component.find('#' + emailField.fieldId).children().length).toBe(2);
	});

	it('A row can have a render prop', () => {
		const _Row = props => <div id={'test2243'}/>;
		const form = {
			fields: [],
			rows: [
				{
					rowId: 'r45',
					render: _Row
				}
			]
		};
		const component = mount(
			<CalderaForm
				form={form}
				onBlur={onBlur}
				onChange={onChange}
			/>
		);
		expect(component.find('#test2243').length).toBe(1);
	});
	it('A column can have a render prop', () => {
		const _Field = props => (
			<input id={'test243'} type={'number'} key={808}/>
		);

		const component = mount(
			<CalderaForm
				form={{
					fields: [],
					rows: [
						{
							rowId: 'r1',
							columns: [
								{
									render: _Field,
									width: '1/2',
									columnId: '1aaaaa'
								},
								{
									fields: [textField],
									width: '1/4',
									columnId: '1b'
								}
							]
						}
					]
				}}
				onBlur={onBlur}
				onChange={onChange}
			/>
		);
		expect(component.find('#test243').length).toBe(1);
	});

	it('A field can have a render prop', () => {
		const _Field = props => <input id={'test808'} type={'number'}/>;

		const field = {
			...checkboxField,
			render: _Field,
			key: 800000
		};

		const form = {
			fields: [checkboxField,textField],
			rows: [
				{
					rowId: 'r1',
					columns: [
						{
							fields: [field, checkboxField],
							width: '1/2',
							columnId: '1aaaaa'
						},
						{
							fields: [textField],
							width: '1/4',
							columnId: '1b'
						}
					]
				}
			]
		};
		const component = mount(
			<CalderaForm
				form={form}
				onBlur={onBlur}
				onChange={onChange}
			/>
		);
		expect(component.find('#test808').length).toBe(1);
		expect(component.find('#test808').prop('type')).toBe('number');
	});
});

describe('Updates ', () => {
	let onSubmit = jest.fn();
	beforeEach(() => {
		onSubmit = jest.fn();
	});

	const form = {
		rows: [
			{
				rowId: 'r1',
				columns: [
					{
						fields: [emailField.fieldId],
						width: '1/2',
						columnId: '1aaaaa'
					},
					{
						fields: [textField.fieldId],
						width: '1/4',
						columnId: '1b'
					}
				]
			}
		],
		fields: [
			emailField,
			textField
		]
	};


	it.skip( 'hides a field when it should ', () => {
		const _form = {
			...form,
			conditionals :[
				{
					type: 'hide',
					rule: createFieldRule( 'is', emailField.fieldId, 'hide' ),
					fields: [
						textField.fieldId
					]
				},

			]
		};

		const component = mount(
			<CalderaForm
				form={form}
				onSubmit={onSubmit}
			/>
		);
		const value = 'hide';
		const event = {target: {value}};
		component
			.find('#emailFieldId')
			.find('input')
			.simulate('change', event);
		expect(
			component
				.find('#firstName')
				.find('input')
				.length
		).toEqual(0);
	});

});
