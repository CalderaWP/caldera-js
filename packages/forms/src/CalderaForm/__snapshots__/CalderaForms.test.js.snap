// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Caldera Forms Forms 1`] = `
<div>
  <form
    onReset={[Function]}
    onSubmit={[Function]}
  >
    <div
      className="sc-bdVaJa caldera-row caldera-form-row sc-bwzfXH pjOvZ"
      id={undefined}
    >
      <div
        className="caldera-column sc-bdVaJa iHZvIS"
        id={undefined}
        width={undefined}
      >
        <div
          className="caldera-field-wrapper caldera-field-wrapper-text"
        >
          <div
            className="components-base-control caldera-field caldera-field-email"
          >
            <div
              className="components-base-control__field"
            >
              <label
                className="components-base-control__label"
                htmlFor="inspector-text-control-0"
              >
                Email Field Label
              </label>
              <input
                aria-describedby="inspector-text-control-0__help"
                className="components-text-control__input"
                id="emailFieldId"
                multiple={true}
                onBlur={[Function]}
                onChange={[Function]}
                placeholder={undefined}
                type="email"
                value={undefined}
              />
            </div>
            <p
              className="components-base-control__help"
              id="inspector-text-control-0__help"
            >
              Email field description
            </p>
          </div>
        </div>
      </div>
      <div
        className="caldera-column sc-bdVaJa iHZvIS"
        id={undefined}
        width={undefined}
      >
        <div
          className="caldera-field-wrapper caldera-field-wrapper-text"
        >
          <div
            className="components-base-control caldera-field caldera-field-text"
          >
            <div
              className="components-base-control__field"
            >
              <label
                className="components-base-control__label"
                htmlFor="inspector-text-control-1"
              >
                First Name
              </label>
              <input
                aria-describedby="inspector-text-control-1__help"
                className="components-text-control__input"
                id="firstName"
                onBlur={[Function]}
                onChange={[Function]}
                placeholder={undefined}
                type="text"
                value={undefined}
              />
            </div>
            <p
              className="components-base-control__help"
              id="inspector-text-control-1__help"
            >
              your first name
            </p>
          </div>
        </div>
      </div>
    </div>
    <input
      className="caldera-forms-submit"
      disabled={false}
      type="submit"
      value="Click Button"
    />
  </form>
</div>
`;
